xblock/mixins.py

```diff
@@ -610,3 +611,19 @@ def has_support(self, view, functionality):
             True or False
         """
         return hasattr(view, "_supports") and functionality in view._supports  # pylint: disable=protected-access
+
+
+class CompletableXBlockMixin(object):
+    has_custom_completion = True
+    completion_method = 'completable'
+
+    def emit_completion(self, completion_ratio):
```

Comment:
I think this method should verify that has_custom_completion is `True` and completion_method is `'completable'`.  The value is set here, but I can imagine a scenario where a subclass wants to fall back to default completion calculation.  If that subclass sets `has_custom_completion = False`, it will fall back to one of the default completion implementations, but also emit its own completion events, which will lead to unexpected behavior.

A test that emit_completion exits silently in those cases would be useful.

Reply:
@jcdyer I don't think it should exit silently - this mixin doesn't have anything useful besides this method - so falling back to default should be performed by not mixing in this class.

Reply:
If the mixin is not a direct superclass, butfurther up the inheritance chain, that won't be an option.  Even though it makes the mixin effectively a noop, that is still the behavior of least surprise and last frustration. The only other options are to cause an error, which will just irritate some poor developer, or to emit the completion anyway, which is exactly the opposite behavior you'd expect when setting has_custom_completion to False.

Reply:
@jcdyer Let me clarify something - we're discussing behavior of custom completion method _when it is called_ on non-custom-compketion block. To me, it is an outright programming error and should be indicated to programmer assoon as possible.

So to me NoCustomCompletion.emit_completion()  is programming error. I realize this mixing can be inherited from some other block, but I can't see a scenario where emit_completion would be rightfully called on non-custom-compketion completion block. If you could describe such scenario I'll gladly revoke this objection. 

In any case, if you really strongly believe silently doing nothing is preferred over raising an error and taking into account you'll be doing majority of work on this feature I'll just stick with your preferred way. However I'm not (yet) convinced it is an optimal long-term solution.